src/compiler/parser.ts

```diff
@@ -2011,6 +2016,34 @@ module ts {
             return parseInitializer(/*inParameter*/ true);
         }
 
+        function parseTypePredicate(signature: SignatureDeclaration) {

```

Comment:
I would not do things this way.  Instead, just augment parseType to work as it does today.  Except that if you parse out a type that is just an identifier, and you see "is" as the current ```token```, then instead parse out a TypePredicateNode.  Then, in the checker if you ever see a TypePredicateNode *not* as the return type of an appropriate signature, then you error.